torch/distributed/_shard/sharded_tensor/partial_tensor.py

```diff
@@ -0,0 +1,154 @@
+                shards_metadata=[
+                    ShardMetadata(shard_offsets=[0], shard_sizes=[2], placement=rank:0/cuda:0),
+                    ShardMetadata(shard_offsets=[2], shard_sizes=[2], placement=rank:1/cuda:1)],
+                size=torch.Size([4])
+        )
+        >>> complete_tensor.local_tensor()
+        tensor([3, 4]) # Rank 0
+        tensor([5, 6]) # Rank 1
+
+        >>> # All tensors below are of torch.cfloat type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([1, 2]) + 2 * rank
+        >>> tensor = torch.cat([tensor, tensor + 2])
+        >>> tensor
+        tensor([1, 2, 3, 4]) # Rank 0
+        tensor([3, 4, 5, 6]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor)
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        ShardedTensor(
+            ShardedTensorMetadata(
+                shards_metadata=[
+                    ShardMetadata(shard_offsets=[0], shard_sizes=[2], placement=rank:0/cuda:0),
+                    ShardMetadata(shard_offsets=[2], shard_sizes=[2], placement=rank:1/cuda:1)],
+                size=torch.Size([4])
+        )
+        >>> complete_tensor.local_tensor()
+        tensor([4, 6]) # Rank 0
+        tensor([8, 10]) # Rank 1
+    """
+
+    def __init__(
+        self, local_shard, process_group=None, reduce_op=distributed_c10d.ReduceOp.SUM
+    ):
+        self.local_shard = local_shard
+        self.process_group = (
+            process_group
+            if process_group
+            else dist.distributed_c10d._get_default_group()
+        )
+        self.reduce_op = reduce_op
+
+    def __post_init__(self):
+        if not isinstance(self.local_shard, torch.Tensor):
+            raise ValueError("local_shard needs to be a Tensor.")
+        if not isinstance(self.reduce_op, distributed_c10d.ReduceOp):
+            raise ValueError(
+                "reduce_op needs to be a member of distributed_c10d.ReduceOp."
+            )
+
+    def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
+        """
+        The reshard happens in two steps logically:
+
+        1. Aggregate all the shards of the partial tensor.
+        2. Shard this tensor according to the provided spec.
+
+        In reality, for the sake of performance, we consolidate all partial tensors
+        across multiple ranks and covert to a sharded tensor in one step.
+
+        Args:
+            resharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+                The specification describing how we reshard the aggregated local result.
+
+        Returns:
+            A :class:`ShardedTensor` filled with local aggregated result.
+        """
+        if not isinstance(resharding_spec, shard_spec.ChunkShardingSpec):
+            raise NotImplementedError("Only ChunkShardingSpec supported for reshard.")
+        if self.local_shard.is_complex():
```

Comment:
question: why we don't support complex tensor on resharding? I feel sharding/resharding should behave the same even with complex tensor, is it because reduce_scatter don't support it?

Reply:
Yes, it's because `reduce_scatter` does not support complex tensor.